lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +434,1140 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+                                          { Cond->getType() },
+                                          false);
+      auto OpaqueTrueCallee = M.getOrInsertFunction("opaque_true", FTy, AL);
+
+      Builder.SetInsertPoint(Term);
+      auto *RetVal = Builder.CreateCall(OpaqueTrueCallee, { Cond });
+
+      if (isa<BranchInst>(Term))
+        cast<BranchInst>(Term)->setCondition(RetVal);
+      else
+        cast<SwitchInst>(Term)->setCondition(RetVal);
+    }
+  }
+
+  // Drop refs, no longer needed.
+  AnyPCCloned = UnexpectedPCCloned = nullptr;
+
+  // Store the values of the 4 PCs in order to have them constant-folded away.
+  for (auto &BB : *OutlinedFunction) {
+    for (auto &I : BB) {
+      if (auto *Call = getCallTo(&I, "newpc")) {
+        MetaAddress NewPC = GCBI::getPCFromNewPC(Call);
+        Builder.SetInsertPoint(Call);
+        PCH->setPC(Builder, NewPC);
+      }
+    }
+  }
+
+  // Some LLVM passes used later in the pipeline scan for cut-offs, meaning that
+  // further computation may not be done when they are reached; making some
+  // optimizations opportunities missed. Hence, we set the involved thresholds
+  // (e.g., the maximum value that MemorySSA uses to take into account
+  // stores/phis) to have initial unbounded value.
+  auto getUOpt = getOption<unsigned>;
+  StringMap<cl::Option *> &Options = cl::getRegisteredOptions();
+  static constexpr const char *MemSSALimit = "memssa-check-limit";
+  static constexpr const char *MemDepBlockLimit = "memdep-block-scan-limit";
+
+  auto MemSSALimitDefault = getUOpt(Options, MemSSALimit)->getValue();
+  auto MemDepBlockLimitDefault = getUOpt(Options, MemDepBlockLimit)->getValue();
+
+  getUOpt(Options, MemSSALimit)->setInitialValue(UINT_MAX);
+  getUOpt(Options, MemDepBlockLimit)->setInitialValue(UINT_MAX);
+
+  // Run optimization passes over the disposable function.
+  {
+    FunctionPassManager FPM;
+
+    // First stage: simplify the IR, promote the CSVs to local variables,
+    // compute subexpressions elimination and resolve redundant expressions in
+    // order to compute the stack height.
+    FPM.addPass(RemoveNewPCCallsPass());
+    FPM.addPass(RemoveHelperCallsPass());
+    FPM.addPass(PromoteGlobalToLocalPass());
+    FPM.addPass(SimplifyCFGPass());
+    FPM.addPass(SROA());
+    FPM.addPass(EarlyCSEPass(true));
+    FPM.addPass(JumpThreadingPass());
+    FPM.addPass(UnreachableBlockElimPass());
+    FPM.addPass(InstCombinePass(true));
+    FPM.addPass(EarlyCSEPass(true));
+    FPM.addPass(SimplifyCFGPass());
+    FPM.addPass(MergedLoadStoreMotionPass());
+    FPM.addPass(GVN());
+
+    // Second stage: add alias analysis info and canonicalize i2p + add into
+    // geps. Since the IR may change remarkably, another round of passes is
+    // necessary to take more optimization opportunities.
+    FPM.addPass(SegregateDirectStackAccessesPass());
+    FPM.addPass(EarlyCSEPass(true));
+    FPM.addPass(InstCombinePass(true));
+    FPM.addPass(GVN());
```

Comment:
This is a quite delicate pipeline... I'm really scared to touch it!
We should spend some time trying to remove an analysis, see what breaks, create a small test out of it and repeat.
Please leave this comment here, no need to do this now.

Reply:
Sounds reasonable.